### PR TITLE
Remove 2 cases of superfluous neutralizer taps and add unit tests for key overrides

### PR DESCRIPTION
**Pull request by precondition**
## Description

Inspired by [sigprof's comment](https://github.com/qmk/qmk_firmware/pull/22209#issuecomment-1749535319) in #22209, I made a few code additions to `quantum/process_keycode/process_key_overrides.c` with negligible increase in firmware size  (my AVR keymap with 3 key overrides created for testing only increased by 14 bytes) that aim to remove some superfluous taps of the `DUMMY_MOD_NEUTRALIZER_KEYCODE`.

There are two main use cases aimed by this pull request:
1. The replacement keycode is a modded keycode (covered by test `do_not_neutralize_if_replacement_uses_any_mods`)
1.1. The replacement keycode's mods are the same as the trigger mods (covered by test `do_not_neutralize_if_replacement_uses_same_mods`)
2. The trigger mods are not suppressed (covered by test `do_not_neutralize_if_no_mods_are_suppressed`)

While I was at it, I added a few more unit tests related to the interplay between key overrides and `DUMMY_MOD_NEUTRALIZER_KEYCODE`, which includes a unit test to illustrate issue #22208 called `modifier_is_suppressed_and_neutralized`.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [ ] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: tests/key_overrides/mod_neutralizer/test_key_overrides.cpp [hunk 1]
```diff
@@ -0,0 +1,432 @@
...
+    key_h.release();
+    EXPECT_REPORT(driver, (KC_LEFT_GUI));
+    run_one_scan_loop();
+    VERIFY_AND_CLEAR(driver);
+
+    key_lgui.release();
+    // The two following reports should also be sent but they're not.
+    // See qmk_firmware issue #22208.
+    // EXPECT_REPORT(driver, (KC_LEFT_GUI, DUMMY_MOD_NEUTRALIZER_KEYCODE));
+    // EXPECT_REPORT(driver, (KC_LEFT_GUI));
```
Comment (precondition) [inline tests/key_overrides/mod_neutralizer/test_key_overrides.cpp:99]: I wasn't sure on what was the best way to add those “wishful” assertions. I know there are (were?) other “wishful” unit tests that highlighted a current bug and wouldn't pass that were configured to be skipped but I didn't like the idea of skipping this whole test due to a single part of it.

**Comment by precondition** [comment]
I don't think this is stale, I'm just awaiting review.

**Review by drashna — APPROVED** [review]

File: quantum/process_keycode/process_key_override.c [hunk 2]
```diff
@@ -325,12 +325,19 @@ static bool try_activating_override(const uint16_t keycode, const uint8_t layer,
...
-        // Send a dummy keycode before unregistering the modifier(s)
-        // so that suppressing the modifier(s) doesn't falsely get interpreted
-        // by the host OS as a tap of a modifier key.
-        // For example, unintended activations of the start menu on Windows when
-        // using a GUI+<kc> key override with suppressed mods.
-        neutralize_flashing_modifiers(active_mods);
+        // If the replacement is a modded keycode, the activation of those mods
+        // will happen in the same keyboard report as the deactivation of the
+        // suppressed mods, so there is no “flashing modifiers” problem.
+        if (!QK_MODS_GET_MODS(override->replacement)) {
```
Comment (zvecr) [inline quantum/process_keycode/process_key_override.c:331]: Is calling `QK_MODS_GET_MODS` here safe without a previous call to check range with `IS_QK_MODS`?

**Comment by precondition** [comment]
Intriguingly, the unit tests I wrote no longer pass now, despite not changing my code. Something in the key processing pipeline elsewhere must have changed :thinking:

**Pull request closed without merging**